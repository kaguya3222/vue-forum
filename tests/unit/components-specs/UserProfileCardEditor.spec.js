import { mount, createLocalVue } from "@vue/test-utils";
import UserProfileCardEditor from "../../../src/components/UserProfileCardEditor";
import usersGetters from "@/store/modules/users/getters";
import Vuex from "vuex";
import mockedRootStore from "../mocks/mockedRootStore";

const localVue = createLocalVue();
const state = mockedRootStore.state;
localVue.use(Vuex);

describe("UserProfileCardEditor", () => {
  const store = new Vuex.Store({
    modules: {
      users: {
        state: { ...state.users },
        namespaced: true,
        getters: { ...usersGetters }
      }
    }
  });
  const $router = {
    push: jest.fn()
  };
  const users = mockedRootStore.state.users.items;
  const user = Object.values(users)[0];
  const wrapper = mount(UserProfileCardEditor, {
    propsData: {
      user
    },
    mocks: {
      $router
    },
    localVue,
    store
  });
  wrapper.vm.$store.dispatch = jest.fn();
  test("Correctly shows form for editing user data", () => {
    expect(wrapper).toMatchSnapshot();
  });
  test("Calls 'save' method when 'Edit' button is clicked", () => {
    const saveMethodSpy = jest.spyOn(wrapper.vm, "save");
    const editButton = wrapper.find(".btn-blue");
    editButton.trigger("click");
    expect(saveMethodSpy).toHaveBeenCalled();
  });
  test("Calls 'cancel' method when 'Cancel' button is clicked", () => {
    const cancelMethodSpy = jest.spyOn(wrapper.vm, "cancel");
    const cancelButton = wrapper.find(".btn-ghost");
    cancelButton.trigger("click");
    expect(cancelMethodSpy).toHaveBeenCalled();
  });
  describe("'save' method", () => {
    beforeEach(() => {
      const editButton = wrapper.find(".btn-blue");
      editButton.trigger("click");
    });
    test("dispatches 'updateUser' action", () => {
      expect(wrapper.vm.$store.dispatch).toHaveBeenCalledWith(
        "updateUser",
        expect.any(Object)
      );
    });
    test("Moves user to 'Profile' route", () => {
      expect(wrapper.vm.$router.push).toHaveBeenCalledWith({ name: "Profile" });
    });
  });
  describe("'cancel' method", () => {
    test("Moves user to 'Profile' route", () => {
      const cancelButton = wrapper.find(".btn-ghost");
      cancelButton.trigger("click");
      expect(wrapper.vm.$router.push).toHaveBeenCalledWith({ name: "Profile" });
    });
  });
});
