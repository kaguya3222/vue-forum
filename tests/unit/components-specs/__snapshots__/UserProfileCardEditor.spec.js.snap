// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserProfileCardEditor Correctly shows form for editing user data 1`] = `
<div class="col-3 push-top">
  <div class="profile-card">
    <p class="text-center"><img src="https://avatars3.githubusercontent.com/u/2327556?v=4&amp;s=460" alt="" class="avatar-xlarge img-update"></p>
    <div class="form-group"><input type="text" placeholder="Username" class="form-input text-lead text-bold"></div>
    <div class="form-group"><input type="text" placeholder="Full Name" class="form-input text-lead"></div>
    <div class="form-group"><label for="user_bio">Bio</label> <textarea id="user_bio" placeholder="Write a few words about yourself." class="form-input"></textarea></div>
    <div class="stats"><span>0 posts</span> <span>0 threads</span></div>
    <hr>
    <div class="form-group"><label for="user_website" class="form-label">Website</label> <input autocomplete="off" id="user_website" class="form-input"></div>
    <div class="form-group"><label for="user_email" class="form-label">Email</label> <input autocomplete="off" id="user_email" class="form-input"></div>
    <div class="form-group"><label for="user_location" class="form-label">Location</label> <input autocomplete="off" id="user_location" class="form-input"></div>
    <div class="btn-group space-between"><button class="btn-ghost">Cancel</button> <button type="submit" class="btn-blue">
        Save
      </button></div>
  </div>
  <p class="text-xsmall text-faded text-center">
    Member since june 2003, last visited 4 hours ago
  </p>
</div>
`;
